inference: inter-procedural conditional constraint back-propagation

aviatesk opened the pull request:
This PR propagates `Conditional`s inter-procedurally when a
`Conditional` at return site imposes a constraint on the call arguments.
When inference exits local frame and the return type is annotated as
`Conditional`, it will be converted into `InterConditional` object,
which is implemented in `Core` and can be directly put into the global
cache. Finally after going back to caller frame, `InterConditional` will
be re-converted into `Conditional` in the context of the caller frame.

 ## improvements 
 
So now some simple "is-wrapper" functions will propagate its constraint
as expected, e.g.:
```julia
isaint(a) = isa(a, Int)
@test Base.return_types((Any,)) do a
    isaint(a) && return a # a::Int
    return 0
end == Any[Int]

isaint2(::Any) = false
isaint2(::Int) = true
@test Base.return_types((Any,)) do a
    isaint2(a) && return a # a::Int
    return 0
end == Any[Int]

function isa_int_or_float64(a)
    isa(a, Int) && return true
    isa(a, Float64) && return true
    return false
end
@test Base.return_types((Any,)) do a
    isa_int_or_float64(a) && return a # a::Union{Float64,Int}
    0
end == Any[Union{Float64,Int}]
```

(and now we don't need something like #38636)

 ## benchmarks

A compile time comparison:
> on the current master (82d79ce18f88923c14d322b70699da43a72e6b32)
```
Sysimage built. Summary:
Total ───────  55.295376 seconds
Base: ───────  23.359226 seconds 42.2444%
Stdlibs: ────  31.934773 seconds 57.7531%
    JULIA usr/lib/julia/sys-o.a
Generating REPL precompile statements... 29/29
Executing precompile statements... 1283/1283
Precompilation complete. Summary:
Total ───────  91.129162 seconds
Generation ──  68.800937 seconds 75.4983%
Execution ───  22.328225 seconds 24.5017%
    LINK usr/lib/julia/sys.dylib
```

> on this PR (37e279bce7136e48f159811641b68143412c3881)
```
Sysimage built. Summary:
Total ───────  51.694730 seconds
Base: ───────  21.943914 seconds 42.449%
Stdlibs: ────  29.748987 seconds 57.5474%
    JULIA usr/lib/julia/sys-o.a
Generating REPL precompile statements... 29/29
Executing precompile statements... 1357/1357
Precompilation complete. Summary:
Total ───────  88.956226 seconds
Generation ──  67.077710 seconds 75.4053%
Execution ───  21.878515 seconds 24.5947%
    LINK usr/lib/julia/sys.dylib
```

Here is a sample code that benefits from this PR:
```julia
function summer(ary)
    r = 0
    for a in ary
        if ispositive(a)
            r += a
        end
    end
    r
end

ispositive(a) = isa(a, Int) && a > 0

ary = Any[]
for _ in 1:100_000
    if rand(Bool)
        push!(ary, rand(-100:100))
    elseif rand(Bool)
        push!(ary, rand('a':'z'))
    else
        push!(ary, nothing)
    end
end

using BenchmarkTools
@btime summer($(ary))
```

> on the current master (82d79ce18f88923c14d322b70699da43a72e6b32)
```
❯ julia summer.jl
  1.214 ms (24923 allocations: 389.42 KiB)
```

> on this PR (37e279bce7136e48f159811641b68143412c3881)
```
❯ julia summer.jl
  421.223 μs (0 allocations: 0 bytes)
```

 ## caveats
  
Within the `Conditional`/`InterConditional` framework, only a single
constraint can be back-propagated inter-procedurally. This PR implements
a naive heuristic to "pick up" a constraint to be propagated when a
return type is a boolean. The heuristic may fail to select an
"interesting" constraint in some cases. For example, we may expect
`::Expr` constraint to be imposed on the first argument of
`Meta.isexpr`, but the current heuristic ends up picking up a constraint
on the second argument (i.e. `ex.head === head`).
```julia
isexpr(@nospecialize(ex), head::Symbol) = isa(ex, Expr) && ex.head === head

@test_broken Base.return_types((Any,)) do x
    Meta.isexpr(x, :call) && return x # x::Expr, ideally
    return nothing
end == Any[Union{Nothing,Expr}]
```

I think We can get rid of this limitation by extending `Conditional` and
`InterConditional`
so that they can convey multiple constraints, but I'd like to leave this
as a future work.

EDIT: 3df027a implements the "pick up" logic within a caller
(i.e. within `abstract_call_gf_by_type`), which allows us to choose a
constraint more appropriately, and now the `Meta.isexpr` case is fixed.
Still there is a limitation of multiple conditional constraint
back-propagation; the following broken test case will explain the future
work.
```julia
is_int_and_int(a, b) = isa(a, Int) && isa(b, Int)
@test_broken Base.return_types((Any,Any)) do a, b
    is_int_and_int(a, b) && return a, b # (a::Int, b::Int) ideally, but (a::Any, b::Int)
    0, 0
end == Any[Tuple{Int,Int}]
```

---

- closes #38636
- closes #37342

jpsamaroo commented:
Slightly off-topic, but  #37342 got accidentally linked for automatic closing.

Keno commented on base/compiler/abstractinterpretation.jl:
```diff
@@ -1144,7 +1144,10 @@ function abstract_eval_statement(interp::AbstractInterpreter, @nospecialize(e),
-        t = callinfo.rt
+        rt = callinfo.rt
+        t = isa(rt, InterConditional) ?
+            transform_from_interconditional(rt, ea) :
```
I think this place is too high level for this transformation. It should probably be inside abstract_call or at the site where we're pulling this out of the cache

aviatesk replied:
yeah I agree, I put this into `abstract_call`.

Keno commented on base/compiler/typeinfer.jl:
```diff
@@ -286,28 +286,34 @@ end
         else
+            if isa(res, Conditional)
+                # TODO: put this into its own field ?
+                rettype = transform_to_interconditional(res, length(result.argtypes))
```
Why not put this into `rettype_const`?

aviatesk replied:
It's just I couldn't make `InterConditional` work in `rettype_const` yesterday. Now it works.

Keno commented on base/compiler/typeinfer.jl:
```diff
@@ -759,15 +766,32 @@ function typeinf_edge(interp::AbstractInterpreter, method::Method, @nospecialize
         update_valid_age!(frame, caller)
-        return frame.bestguess, frame.inferred ? mi : nothing
+        bestguess = frame.bestguess
+        if isa(bestguess, Conditional)
```
I'd rather split this out into a descriptive function e.g. `bestguess_to_interprocedural` that gets calls unconditionally and then checks this situation internally.

aviatesk replied:
Yeah good idea, I put `bestguess_to_interprocedural` into `finish(::InferenceState, ::AbstractInterpreter)`.

aviatesk commented:
All tests turned green. Ready for review.

Keno commented on base/compiler/typelattice.jl:
```diff
@@ -45,6 +44,15 @@ struct Conditional
+#     elsetype
+# end
+import Core: InterConditional
+const ConditionalWrapper = Union{Conditional,InterConditional}
```
I'm not sure I like `ConditionalWrapper`. Maybe `AnyConditional`?

aviatesk replied:
renamed

KristofferC commented:
Out of curiosity, any measurements regarding compile times? For example, building the sysimage, time to first plot etc.

Keno commented on base/compiler/typelattice.jl:
```diff
@@ -103,9 +108,13 @@ function issubconditional(a::Conditional, b::Conditional)
     return false
 end
+is_same_conditionals(a::Conditional,      b::Conditional)      = slot_id(a.var) === slot_id(b.var)
+is_same_conditionals(a::Conditional,      b::InterConditional) = slot_id(a.var) === b.slot
```
When would we ever compare a `Conditional` to an `InterConditional`? They seem valid in exactly opposite situations, so it feels like comparing them would be a bug. Am I missing something?

aviatesk replied:
When constant propagation happens, the previous `InterConditional` result can be compared to the new `Conditional`:
https://github.com/JuliaLang/julia/blob/cec310eedb5c263d57623840b635d8f1dbf43402/base/compiler/abstractinterpretation.jl#L167

Keno replied:
I think we should structure the code so this does not happen. Everything should either be an interprocedurally safe type or not. Doing something like that just seems to ask for trouble. I guess the other option would be to not have `InterConditional` at all and just use `Conditional` everywhere, but if we're splitting the lattice like this, we should keep the comparisons separate.

aviatesk replied:
Ops, sorry it's just my mistake... Your insight is correct, and actually they've never been compared.
I removed those diagonal dispatch signatures.

aviatesk commented:
> Out of curiosity, any measurements regarding compile times? For example, building the sysimage, time to first plot etc.

I'll get back something here later ;) For now I'd say this PR seems to speed up both compilation/execution times a bit.

aviatesk commented:
Here is a compile time comparison:

> on the current master 3853060e54c4995740df1b4e7247bafddd9dbf0c
```
Sysimage built. Summary:
Total ───────  58.692251 seconds 
Base: ───────  21.985050 seconds 37.4582%
Stdlibs: ────  36.705812 seconds 62.5395%
    JULIA usr/lib/julia/sys-o.a
Generating REPL precompile statements... 29/29
Executing precompile statements... 1279/1279
Precompilation complete. Summary:
Total ─────── 114.426189 seconds
Generation ──  90.801799 seconds 79.354%
Execution ───  23.624390 seconds 20.646%
    LINK usr/lib/julia/sys.dylib
```

> this PR cec310eedb5c263d57623840b635d8f1dbf43402
```
Sysimage built. Summary:
Total ───────  53.670412 seconds
Base: ───────  22.107793 seconds 41.1918%
Stdlibs: ────  31.561120 seconds 58.8054%
    JULIA usr/lib/julia/sys-o.a
Generating REPL precompile statements... 29/29
Executing precompile statements... 1300/1300
Precompilation complete. Summary:
Total ─────── 109.051717 seconds
Generation ──  86.042248 seconds 78.9004%
Execution ───  23.009469 seconds 21.0996%
    LINK usr/lib/julia/sys.dylib
```

---

Regarding @Keno's comment
> I guess the other option would be to not have InterConditional at all and just use Conditional everywhere,

I actually tried this option first (you can see [this branch](https://github.com/aviatesk/julia/tree/backprop3)), but using `Conditional` everywhere somehow leads to excessive code generations and significant codegen slow down as shown below:

> using `Conditional` everywhere cebb2713cde4a892e80f14422b7c7566676c8c5e
```
Sysimage built. Summary:
Total ───────  77.697483 seconds
Base: ───────  32.633602 seconds 42.0008%
Stdlibs: ────  45.062015 seconds 57.9967%
    JULIA usr/lib/julia/sys-o.a
Generating REPL precompile statements... 29/29
Executing precompile statements... 1522/1522
Precompilation complete. Summary:
Total ─────── 227.906477 seconds
Generation ── 164.706820 seconds 72.2695%
Execution ───  63.199657 seconds 27.7305%
    LINK usr/lib/julia/sys.dylib
```

aviatesk commented on base/compiler/tfuncs.jl:
```diff
@@ -1648,7 +1648,7 @@ function return_type_tfunc(interp::AbstractInterpreter, argtypes::Vector{Any}, s
                         return Const(Union{})
                     end
-                    rt = abstract_call(interp, nothing, argtypes_vec, sv, -1).rt
+                    rt = widenconditional(abstract_call(interp, nothing, argtypes_vec, sv, -1).rt)
```
Now this change isn't necessary for this PR, but I think `widenconditional` is better here since it may produce more accurate result since it can be widen to `Const`, so let me include this as is.

aviatesk commented on base/compiler/abstractinterpretation.jl:
```diff
@@ -1033,9 +1033,35 @@ function abstract_call(interp::AbstractInterpreter, fargs::Union{Nothing,Vector{
+        # try to convert interprocedural conditional constraint from callee into the constraint
+        # on slots of the current frame
+        i = rt.slot
+        if checkbounds(Bool, ea, i)
```
I believe we don't need this bound check since `InterConditional` only comes from a "valid" call and its slot is supposed to be always within the bound of the call argument (I confirmed tests got passed w/o this).
I added this just for robustness and safety, but I'm not sure whether or not we should remove this. What do you think ?

aviatesk commented:
Okay, I addressed the review comments, and this is good to go, I'd say. Any further comments and reviews ?

aviatesk commented:
bump, can we push this forward ? 
Could any of you have a review on this, @Keno , @JeffBezanson or @vtjnash ?

vtjnash reviewed (approved):
This is a really nice job! I've been following along as I could, but needed to get a chunk of review time. I think with a few minor updates this should be good to merge.

aviatesk commented on base/compiler/typeinfer.jl:
```diff
@@ -412,10 +416,21 @@ function finish(me::InferenceState, interp::AbstractInterpreter)
+function bestguess_to_interprocedural(bestguess::Conditional, nargs::Int)
+    # keep `Conditional` return type only when it constrains any of call argument
+    i = slot_id(bestguess.var)
+    if 1 < i ≤ nargs
```
I excluded `1` because I couldn't come up with any practical example where we can benefit from constraints on a called generic function itself.
(I actually came up with this one
```julia
struct F{T}
    v::T
end
(f::F)(T) = isa(f, F{T})

code_typed((F,); optimize = false) do f
    f(Int) && return f.v # ::Int
    0
end |> first
``` 
but this kind of code seems really useless)

Well at the same time, I don't think there is any disadvantage to have an `InterConditional` on a slot 1, though.

vtjnash replied:
yeah, it's not likely to happen, but seemed feasible to easily include, and that just means it won't typically cost anything

aviatesk commented on base/compiler/abstractinterpretation.jl:
```diff
@@ -1033,9 +1033,35 @@ function abstract_call(interp::AbstractInterpreter, fargs::Union{Nothing,Vector{
+            end
+        end
+        return CallMeta(widenconditional(rt), callinfo.info)
+    else
```
Is it okay to have `@assert`s for this kind of (somewhat performance-critical) code ? 
I have a naive assumption that `throw` calls lead to worse performance.  Maybe only `try/catch` matters in performance aspect, but `throw` doesn't ?

vtjnash replied:
try/catch is slow, but the existence of unreached code here is (nearly) free

aviatesk replied:
okay, let's assert we don't mix-up `Conditional` in inter-procedural context.

Well, `@assert` shouldn't be here, since `abstract_call_builtin` can return `Conditional`s. 
I'd do this assertion in `abstract_call_gf_by_type`): 25c5f151e791b2bd8a0a6c3f11c6e2171504a5cc

aviatesk commented on base/essentials.jl:
```diff
@@ -813,8 +813,7 @@ const missing = Missing()
 """
-ismissing(::Any) = false
-ismissing(::Missing) = true
+ismissing(x) = x === missing
```
Hmm, I'm not sure 

> This seems somewhat harder to infer and cache, though we want it to be equivalent.

Could you please elaborate this ?

I included this change (as I did for `isnothing`) considering we can benefit from this with some code like below
```julia
code_typed((Vector{Union{Missing,Int}},); optimize = false) do v
    s = 0.
    for a in v
        ismissing(a) && continue
        s + sin(a) # type stable, `a::Int`
    end
    s
end |> first
```

vtjnash replied:
Ah, right, because we don't infer it due to the early union-splitting. I think we should? (it's quite similar to the `ispositive` case also in this framework)

aviatesk replied:
yes, that's one of the current limitations. For now `Conditional`s never appear in code like below:
```julia
checker(@nospecialize(_)) = false
checker(::SomeType1) = true
checker(::SomeType2) = true
```

nalimilan replied:
What happens if a package defines `Base.ismissing(::MyCustomMissing) = true`? This is something that could be useful at some point. Would that break optimizations?

aviatesk replied:
The type constraint will be there as far as the number of `ismissing` methods is [≤3](https://github.com/JuliaLang/julia/blob/5c6e21edbfd8f0c7d16ea01c91d1c75c30d4eaa1/base/compiler/types.jl#L119)
```julia
julia> struct MyCustomMissing end

julia> Base.ismissing(::MyCustomMissing) = true

julia> code_typed((Any,)) do a
           ismissing(a) && return a
           0
       end
1-element Vector{Any}:
 CodeInfo(
1 ─ %1 = (isa)(a, MyCustomMissing)::Bool
└──      goto #3 if not %1
2 ─      goto #4
3 ─ %4 = Main.ismissing(a)::Bool
└──      goto #4
4 ┄ %6 = φ (#2 => true, #3 => %4)::Bool
└──      goto #6 if not %6
5 ─ %8 = π (a, Union{Missing, MyCustomMissing})
└──      return %8
6 ─      return 0
) => Union{Missing, MyCustomMissing, Int64}
```

aviatesk commented on test/compiler/inference.jl:
```diff
@@ -1719,6 +1719,41 @@ for expr25261 in opt25261[i:end]
+
+    # FIXME: we can't propagate conditional constraints interprocedurally when there're
+    # multiple possible conditions within the callee
+    ispositive(a) = isa(a, Int) && a > 0
```
yes, I _think_ we can change our local inference logic so that we can annotate `Conditional` for a statement (e.g. `%5`) with considering where it comes from (e.g. `%1`), but I didn't try to tackle that in this PR because I felt the changes can be huge.
Maybe we can merge this PR with only introducing `InterConditional` and leave this as a future work (in a separate PR) ?

vtjnash replied:
The change should be pretty simple, though probably doesn't need to be in this PR: if the inferred result is <: Bool, then it is actually Conditional for the entire VarTable `s[pc][1:nargs]`. The annoying part then is that InterConditional can only represent exactly one of those args, so you may have to just pick the first interesting one. (conceptually, this is a also a generalizable observation on what Conditional represents: either it's an alternative VarTable based on a particular value comparison result—or the main VarTable `s` is a Conditional on the program state)

aviatesk replied:
Okay, with the following diff, we can back-propagate constraints in cases like `ispositive` and also some cases like 
the original definitions of `isnothing` and `ismissing` (it looks dirty, though):
```julia
diff --git a/base/compiler/abstractinterpretation.jl b/base/compiler/abstractinterpretation.jl
index 98b8b1c9c9..6a09b74b3e 100644
--- a/base/compiler/abstractinterpretation.jl
+++ b/base/compiler/abstractinterpretation.jl
@@ -1300,6 +1300,7 @@ function typeinf_local(interp::AbstractInterpreter, frame::InferenceState)
     W = frame.ip
     s = frame.stmt_types
     n = frame.nstmts
+    nargs = frame.nargs
     while frame.pc´´ <= n
         # make progress on the active ip set
         local pc::Int = frame.pc´´ # current program-counter
@@ -1362,6 +1363,25 @@ function typeinf_local(interp::AbstractInterpreter, frame::InferenceState)
             elseif isa(stmt, ReturnNode)
                 pc´ = n + 1
                 rt = abstract_eval_value(interp, stmt.val, s[pc], frame)
+                if isa(rt, Conditional) && !(1 ≤ slot_id(rt.var) ≤ nargs)
+                    # discard this `Conditional` imposed on non-call arguments,
+                    # since it's not interesting in inter-procedural context;
+                    # other ReturnNode statement may give constraints on any of call argument
+                    rt = widenconditional(rt)
+                end
+                if rt !== Bottom && !isa(rt, Conditional) && rt ⊑ Bool
+                    if isa(frame.bestguess, Conditional)
+                        # if the bestguess so far is already `Conditional`, try to convert
+                        # this `rt` into `Conditional` on the slot to avoid overapproximation
+                        # due to conflict of different slots
+                        rt = boolean_rt_to_conditional(rt, changes, slot_id(frame.bestguess.var))
+                    elseif nargs ≥ 1
+                        # pick up the first "interesting" slot, convert `rt` to its `Conditional`
+                        # TODO: this is very naive heuristic, ideally we want something like
+                        # alternative `VarTable` that keeps conditional constraints on all the variables
+                        rt = boolean_rt_to_conditional(rt, changes, nargs == 1 ? 1 : 2)
+                    end
+                end
                 if !isa(rt, Const) &&
                    !isa(rt, Type) &&
                    !isa(rt, PartialStruct) &&
@@ -1478,6 +1498,21 @@ function typeinf_local(interp::AbstractInterpreter, frame::InferenceState)
     nothing
 end
 
+function boolean_rt_to_conditional(@nospecialize(rt), state::VarTable, slot_id::Int)
+    varstate = state[slot_id]::VarState
+    typ = widenconditional(varstate.typ) # avoid nested conditional
+    if isa(rt, Const)
+        if rt.val === true
+            return Conditional(SlotNumber(slot_id), typ, Bottom)
+        elseif rt.val === false
+            return Conditional(SlotNumber(slot_id), Bottom, typ)
+        end
+    elseif rt === Bool
+        return Conditional(SlotNumber(slot_id), typ, typ)
+    end
+    return rt # when `rt` is Union{}, etc.
+end
+
 # make as much progress on `frame` as possible (by handling cycles)
 function typeinf_nocycle(interp::AbstractInterpreter, frame::InferenceState)
     typeinf_local(interp, frame)
```

But I couldn't succeed to bootstrap with this (with `fatal error in type inference (type bound)` error).. 
Do you have any pointer about where to fix ?

vtjnash replied:
On closer inspection, I think you're missing `bestguess_to_interprocedural` on some of the uses of `bestguess`. While you could grep each usage, I think it might be better to consolidate it here: (I thought the new assert should have caught that)

```diff
diff --git a/base/compiler/abstractinterpretation.jl b/base/compiler/abstractinterpretation.jl
index 98b8b1c9c9..c76c61e89b 100644
--- a/base/compiler/abstractinterpretation.jl
+++ b/base/compiler/abstractinterpretation.jl
@@ -1362,14 +1362,11 @@ function typeinf_local(interp::AbstractInterpreter, frame::InferenceState)
             elseif isa(stmt, ReturnNode)
                 pc´ = n + 1
                 rt = abstract_eval_value(interp, stmt.val, s[pc], frame)
-                if !isa(rt, Const) &&
-                   !isa(rt, Type) &&
-                   !isa(rt, PartialStruct) &&
-                   !isa(rt, Conditional)
-                    # only propagate information we know we can store
-                    # and is valid inter-procedurally
-                    # some of them will further be transformed by `bestguess_to_interprocedural`
-                    # in `finish(::InferenceState, ::AbstractInterpreter)`
+                # only propagate information we know we can store
+                # and is valid inter-procedurally
+                if isa(rt, Conditional)
+                    rt = bestguess_to_interprocedural(rt, frame.nargs)
+                elseif !isa(rt, Const) && !isa(rt, Type) && !isa(rt, PartialStruct)
                     rt = widenconst(rt)
                 end
                 if tchanged(rt, frame.bestguess)
diff --git a/base/compiler/typeinfer.jl b/base/compiler/typeinfer.jl
index 8a2d22b252..6cbe1fb5f0 100644
--- a/base/compiler/typeinfer.jl
+++ b/base/compiler/typeinfer.jl
@@ -416,11 +416,10 @@ function finish(me::InferenceState, interp::AbstractInterpreter)
         me.result.src = OptimizationState(me, OptimizationParams(interp), interp)
     end
     me.result.valid_worlds = me.valid_worlds
-    me.result.result = me.bestguess = bestguess_to_interprocedural(me.bestguess, me.nargs)
+    me.result.result = me.bestguess
     nothing
 end
 
-bestguess_to_interprocedural(@nospecialize(bestguess), _::Int) = bestguess
 function bestguess_to_interprocedural(bestguess::Conditional, nargs::Int)
     # convert `Conditional` into `InterConditional` only when it imposes constrains on
     # any of call argument
@@ -787,7 +786,7 @@ function typeinf_edge(interp::AbstractInterpreter, method::Method, @nospecialize
     # return the current knowledge about this cycle
     frame = frame::InferenceState
     update_valid_age!(frame, caller)
-    return bestguess_to_interprocedural(frame.bestguess, frame.nargs), nothing
+    return frame.bestguess, nothing
 end
 
 #### entry points for inferring a MethodInstance given a type signature ####
```

vtjnash replied:
Okay, dug deeper and found it is due to missing handling of Vararg expansion in this PR. Here's a reduced repro. Note how the `x...` turned into a `x::Tuple{Symbol}` type description (which will fail at runtime, because actually we know that `x::Symbol` there)

```julia
julia> code_typed(mapfilter, (typeof((x...) -> !isa(x, Tuple{Integer})), typeof(Base.delete!), Base.Set{Union{Int64, Symbol}}, Base.Set{Union{Int64, Symbol}}) , optimize=false)                       
1-element Vector{Any}:                                                                             
 CodeInfo(                                                                                         
    @ REPL[8]:2 within `mapfilter'                                                                                                                                                                     
1 ─ %1  = a::Set{Union{Int64, Symbol}}                                                                                                                                                                 
│         (@_6 = Base.iterate(%1))::Union{Nothing, Tuple{Union{Int64, Symbol}, Int64}}             
│   %3  = (@_6 === nothing)::Bool                                                                  
│   %4  = Base.not_int(%3)::Bool                                                                   
└──       goto #6 if not %4                                                                        
2 ┄ %6  = @_6::Tuple{Union{Int64, Symbol}, Int64}::Tuple{Union{Int64, Symbol}, Int64}              
│         (x = Core.getfield(%6, 1))::Union{Int64, Symbol}                                                                                                                                             
│   %8  = Core.getfield(%6, 2)::Int64                                                              
│   @ REPL[8]:3 within `mapfilter'                                                                 
│   %9  = (pred)(x)::Bool                                                                          
└──       goto #4 if not %9                                                                        
3 ─       (f)(b, x::Tuple{Symbol})::Any                                                                                                                                                                
└──       goto #4                                                                                  
4 ┄       (@_6 = Base.iterate(%1, %8))::Union{Nothing, Tuple{Union{Int64, Symbol}, Int64}}         
│   %14 = (@_6 === nothing)::Bool                                                                                                                                                                      
│   %15 = Base.not_int(%14)::Bool                                                                  
└──       goto #6 if not %15                                                                                                                                                                           
5 ─       goto #2                                                                                  
6 ┄       return nothing                                                                           
) => Nothing                              

julia> s = Set{Union{Int,Symbol}}((:a, 1)); mapfilter((x...) -> !isa(x, Tuple{Integer}), delete!, s, s)
signal (11): Segmentation fault                                                         
```

(was hoping to be able to explain simply how I found this, but ended up being a bit convoluted. Mostly I ran under `rr` until exit, then ran backwards to `break jl_throw`, then back to `watch` on the value of $rip, then back to `eval_function` a couple times, until I could see the printing of `p jl_(src)` for the correct `p jl_(li)` function. I searched this for "fatal" and saw `Tuple{Int}`, which I could tell was bad, so then it was "just" a matter of guessing different IR inputs until one propagated the type wrong)

aviatesk replied:
Thanks so much for your help !
Unfortunately I'm on macOS and it seems to me that debugging w/ gdb is hard on that platform (rr seems even unavailable ...). 
I'll try to get Linux or docker environment to develop Julia, but for now I'd like to move forward this PR.

12e1e79 enhances this PR by eagerly converting boolean conditions into `(Inter)Conditional`s following your idea.
Now I succeed in bootstrapping but it seems that there're few inference accuracy regressions.

aviatesk replied:
Okay I just fixed the inference accuracy regression; there're some code that make good use of `Const(val::Bool)` to improve inference quality (e.g. [`ifelse` tfunc](https://github.com/JuliaLang/julia/blob/d46198d1bbc19165e45f0b0a7003253ccff0773b/base/compiler/tfuncs.jl#L215-L231)), but now we turn `Const(val::Bool)` return type into `Conditional`, and so we need to respect its constant-ness to keep accuracy.
As for `ifelse` one, I fixed that in fe6accd, but there may be other points that should be tweaked.

Once we've done all of them, I think this PR should be good to go.

vtjnash replied:
Yes, on macOS, lldb is better, but setting up a VM just to run `rr` in linux is also great (Can you tell `rr` is one of my favorite magic tools?)

I think we also might want to stick with using `Const(::Bool)` as the result object when the new type doesn't refine the old type of the Argument (e.g. except when `⊑(s[pc][i], s[1][i])`), then choose later (in `abstract_call_gf_by_type`) to form a tighter bound if union-splitting with `napplicable > 1`. Note, you can examine `fieldtype(sig, n)` to find out the type of the nth argument.

aviatesk commented on base/compiler/abstractinterpretation.jl:
```diff
@@ -1187,9 +1213,9 @@ function abstract_eval_statement(interp::AbstractInterpreter, @nospecialize(e),
-            if isa(at, Const) && isa(at.val, Tuple) && n == length(at.val) &&
-                let t = t, at = at; _all(i->at.val[i] isa fieldtype(t, i), 1:n); end
-                t = Const(ccall(:jl_new_structt, Any, (Any, Any), t, at.val))
+            if isa(at, Const) && (val = at.val; isa(val, Tuple)) && n == length(val) &&
+                let t = t, val = val; _all(i->val[i] isa fieldtype(t, i), 1:n); end
+                t = Const(ccall(:jl_new_structt, Any, (Any, Any), t, val))
```
These changes don't relate to this PR, I included several code quality improvements in f00abf90

aviatesk commented on base/compiler/abstractinterpretation.jl:
```diff
@@ -1312,19 +1340,21 @@ function typeinf_local(interp::AbstractInterpreter, frame::InferenceState)
                     pc´ = l
+                    if isa(condt, Conditional)
+                        changes = conditional_changes(changes, condt.elsetype, condt.var)
+                    end
```
Constant conditions can be `Conditional` and convey an additional constraint, and these changes will propagate it as well as the "general case" `Conditional`s.

E.g. the following example will be improved:
```julia
f(a) = g(a) ? a : nothing
g(a::Int) = true

Base.return_types(f, (Any,)) # ::Int
```

vtjnash replied:
Yes, these really should have an even stronger effect than that and act like a type-assert on the statement, which then constrains `a` on every subsequent statement. We can't handle that general case yet, but for now, it might be more accurate to put this logic into `stupdate!` to occur after every statement, instead of being specific to branches. I'm inclined to say we should omit this particular change, because it is largely unrelated to this PR, and is unable to handle many other cases.

```
f(a) = (g(a); a)
g(a::Int) = true # ::Conditional(SlotNumber(2), Int, Union{}), constant conditional constraint

@test Base.return_types(f, (Any,)) == Any[Int]
```

aviatesk replied:
thanks, I agree with you, let's improve this in higher level (reverted in 6b31e2d)

aviatesk commented on base/compiler/abstractinterpretation.jl:
```diff
@@ -1450,6 +1508,27 @@ function typeinf_local(interp::AbstractInterpreter, frame::InferenceState)
 end
 
+function conditional_changes(changes::VarTable, @nospecialize(typ), var::Slot)
+    if typ ⊑ (changes[slot_id(var)]::VarState).typ
```
The original code was https://github.com/JuliaLang/julia/blob/d46198d1bbc19165e45f0b0a7003253ccff0773b/base/compiler/abstractinterpretation.jl#L1322-L1325

I changed that so that we propagate new conditional type only when it is really worthwhile by checking it's lower in the lattice than the original type.

vtjnash replied:
I mostly assumed that was assured by construction before (since we build these with typesubtract and typeintersect, or explicitly with Any), but makes sense also to be explicit

aviatesk commented on base/meta.jl:
```diff
@@ -73,10 +73,14 @@ julia> Meta.isexpr(ex, :call, 2)
-isexpr(@nospecialize(ex), head::Symbol) = isa(ex, Expr) && ex.head === head
+function isexpr end
+
+# seemingly useless `head′` and `n′` are here just to help inference back-propagate
+# `isa(ex, Expr)` constraints inter-procedurally
+isexpr(@nospecialize(ex), head::Symbol) = isa(ex, Expr) && (head′ = head; ex.head === head′)
 isexpr(@nospecialize(ex), heads) = isa(ex, Expr) && in(ex.head, heads)
-isexpr(@nospecialize(ex), head::Symbol, n::Int) = isa(ex, Expr) && ex.head === head && length(ex.args) == n
-isexpr(@nospecialize(ex), heads, n::Int) = isa(ex, Expr) && in(ex.head, heads) && length(ex.args) == n
+isexpr(@nospecialize(ex), head::Symbol, n::Int) = isa(ex, Expr) && (head′ = head; ex.head === head′) && (n′ = n; length(ex.args) == n′)
+isexpr(@nospecialize(ex), heads, n::Int) = isa(ex, Expr) && in(ex.head, heads) && (n′ = n; length(ex.args) == n′)
```
These `head′` and `n′` tweaks would look really weird, but they help us convey `isa(ex, Expr)` constraint to caller under the current naive "condition pick up" heuristic.

aviatesk commented:
Okay, now this PR is in pretty good shape, I think. Could you review this again, @vtjnash , @Keno ?

There is one failed existing test: https://github.com/JuliaLang/julia/blob/d46198d1bbc19165e45f0b0a7003253ccff0773b/test/compiler/inference.jl#L1255
This is because now `(x)->isdefined(x, :re)` doesn't return `Const(true)` but rather `InterConditional(2, ComplexF64, Union{})`. Can I change/remove this test case ?
(Well, as I posted on `#internal` slack channel, I'm wondering why `Core.Compiler.invoke_api` doesn't return `2` from caches with`PartialStruct` or `InterConditional` `rettype_const`. Is this expected ?)

vtjnash commented on base/meta.jl:
```diff
@@ -73,10 +73,14 @@ julia> Meta.isexpr(ex, :call, 2)
+
+# seemingly useless `head′` and `n′` are here just to help inference back-propagate
+# `isa(ex, Expr)` constraints to the caller inter-procedurally
+isexpr(@nospecialize(ex), head::Symbol) = isa(ex, Expr) && (head′ = head; ex.head === head′)
 isexpr(@nospecialize(ex), heads) = isa(ex, Expr) && in(ex.head, heads)
-isexpr(@nospecialize(ex), head::Symbol, n::Int) = isa(ex, Expr) && ex.head === head && length(ex.args) == n
-isexpr(@nospecialize(ex), heads, n::Int) = isa(ex, Expr) && in(ex.head, heads) && length(ex.args) == n
+isexpr(@nospecialize(ex), head::Symbol, n::Int) = isa(ex, Expr) && (head′ = head; ex.head === head′) && (n′ = n; length(ex.args) == n′)
+isexpr(@nospecialize(ex), heads, n::Int) = isa(ex, Expr) && in(ex.head, heads) && (n′ = n; length(ex.args) == n′)
```
IIUC, I think this problem should be solved now?

vtjnash replied:
Ah, I see, it still prioritizes the test against `head`, though it yields no beneficial information

aviatesk replied:
yeah, maybe this comment will explain to those who're interested in the future:
https://github.com/JuliaLang/julia/blob/b08f86f335a5a878e50a22714ae9233c5877f096/test/compiler/inference.jl#L1789-L1797

vtjnash replied:
I think fixing that is the last piece to solve for this PR (and would also fix the `test_const_return((x)->isdefined(x, :re), Tuple{ComplexF64}, true)` test). Namely, that Conditional results which don't constrain the input argument should be left as Const instead. Then, logic similar to what you wrote to handle ReturnNode, where you pick up the first "interesting" slot, would be needed in `abstract_call_gf_by_type` (when `napplicable > 1`) to form it just in time when it is needed.

ref https://github.com/JuliaLang/julia/pull/38905/files#r552171097

aviatesk replied:
Hmmm, do you mean something like [this](https://github.com/aviatesk/julia/compare/backprop2...aviatesk:backprop4) ?
I'm not sure I'm following your idea, but to me the diff looks a bit complicated and I feel it might be better to keep doing the "pick up interesting slot, covert boolean into conditional" work in just one place, for now.
Could you tell me why we should return `Const(::Bool)` instead of `InterConditional` which contains essential information about the constantness of its boolean value ?

I also wonder that it won't solve the original `isexpr` case even if we stop converting `Const(::Bool)` into `Conditional` in a callee frame.
As far as I understand, the current limitation is that we can't back-propagate multiple conditionals; as shown below,
with the lowering code obtained from the original `isexpr` definition, we will first see `Conditional(SlotNumber(3), Symbol, Symbol)` return type and so we will [try to convert the succeeding boolean return types into conditional on the `SlotNumber(3)`](https://github.com/JuliaLang/julia/blob/4fa55985685d45d8df77286c8c1bb927ca80a968/base/compiler/abstractinterpretation.jl#L1381-L1385), which is less interesting than that on `SlotNumber(2)`. In other word, since `SlotNumber(3)` is a valid argument and we first see it as return type, the current logic ends up missing the constraint of `SlotNumber(2)`.
This problem won't change whatever we try to convert `Const(false)` return type at `%7` into `Conditional`, no ?
```julia
julia> isexpr(@nospecialize(ex), head::Symbol) = isa(ex, Expr) && ex.head === head
isexpr (generic function with 1 method)

julia> code_lowered(isexpr, (Any,Symbol)) |> first
CodeInfo(
1 ─      nothing
│   %2 = ex isa Main.Expr
└──      goto #3 if not %2
2 ─ %4 = Base.getproperty(ex, :head)
│   %5 = %4 === head  # Conditional(SlotNumber(3), Symbol, Symbol)
└──      return %5    # Conditional(SlotNumber(3), Symbol, Symbol) -> Conditional(SlotNumber(3), Symbol, Symbol)
3 ─      return false # Const(false) -> Conditional(SlotNumber(3), Union{}, Symbol)
)
```
The new definition of `isexpr` setups a dummy local variable for `SlotNumber(3)` and prevents the logic to propagate its constraint (because the dummy is not a call argument, the logic will propagate the constraint of `SlotNumber(2)` instead).

I guess the best way to address this would be to setup new lattice element that contains multiple conditional constraints, back-propagate it and update states based with it. Since it will end up refactoring huge part of our lattice implementation, I thought it's better to be done in another PR.

aviatesk replied:
@vtjnash could I get a quick comment on this ? I'd like to know whether I'm understanding your idea or missing something.

vtjnash replied:
Yes, that commit looks about right. I ran into an unrelated bug trying to use it, so let me fix that first (https://github.com/JuliaLang/julia/issues/39218) and I'll push some updates

aviatesk replied:
okay, I'm being pretty busy for day work until this weekend, but will try to help you fix the bug once I find a time.

vtjnash replied:
Finished now in cb07897d87dd2b39019336e93c1d5586465277ef. WDYT?

aviatesk replied:
Thanks, #39221 looked good to me.

I just rebased this PR against the current `master`; I found some [tests with logging macro](https://github.com/aviatesk/julia/blob/18f5f7cef783acd9e00d7174023b1102afeea226/test/corelogging.jl#L97-L112) are failing on this PR somehow, and so will investigate into it, but I guess you may have better insight on this because the blame says you added those tests ?

I also updated [`backprop4`](https://github.com/aviatesk/julia/compare/backprop2...aviatesk:backprop4) branch, which tries to implement [your idea](https://github.com/JuliaLang/julia/pull/38905#discussion_r552353609). How you do want to review it ? Shall I push the commit into this PR, or craft a separate PR ?

aviatesk replied:
Okay I've tracked down the logging macro failures are because of a bug in this PR:
```julia
julia> code_typed((Any,); optimize = false) do x
           Base.CoreLogging.issimple(x) && return x
           nothing
       end |> first
CodeInfo(
1 ─ %1 = Base.CoreLogging::Core.Const(Base.CoreLogging)
│   %2 = Base.getproperty(%1, :issimple)::Core.Const(Base.CoreLogging.issimple)
│   %3 = (%2)(x)::Bool
└──      goto #3 if not %3
2 ─      return x::String
3 ─      return Main.nothing
) => Union{Nothing, String}
```
, and this leads to the wrong code generation of `logmsg`.

aviatesk replied:
So this PR is failing because e1859a1c25a returns `true` for different constant `Conditional`.
```julia
t1, t2 = InterConditional(2, Int, Union{}), InterConditional(2, Symbol, Union{}) # different conditionals
t1 ⊑ t2          # => true
tchanged(t1, t2) # => false
```
, with which we won't update `InterConditional` return type when there're multiple constant `Conditional` returns, e.g.
```julia
f(x) = begin
    isa(x, Symbol) && return true
    isa(x, Int) && return true
    false
end

code_typed((Any,); optimize = false) do x
    f(x) && return x
    nothing
end |> first |> last # => Union{Nothing, Symbol}
```

---

I think we can just fix this by not converting `Const(::Bool)` return type to `InterConditional`. Let me merge `backprop4` branch into this PR, and see how CI goes.

aviatesk replied:
I also added more fix for e1859a1 in d30287f; we need to unwrap same constant conditionals on different slots earlier, otherwise 
```julia
# before
Core.Compiler.tmerge(Core.Compiler.InterConditional(1, Int, Union{}), Core.Compiler.InterConditional(2, String, Union{})) # => Bool

# after
Core.Compiler.tmerge(Core.Compiler.InterConditional(1, Int, Union{}), Core.Compiler.InterConditional(2, String, Union{})) # => Const(true)
```

vtjnash replied:
Ah, sorry, I've reverted that bad commit.

Did you see my link https://github.com/JuliaLang/julia/commit/cb07897d87dd2b39019336e93c1d5586465277ef? That was based on your backprop4 branch, but should finish the implementation of it. You're welcome to take that and squash it into your commits. It should always handle the case you are concerned about there (though I don't currently have the branch checked out to verify for certain).

aviatesk replied:
Ah, sorry I've totally missed cb07897 ! 

The code looks good (rather, amazing) to me; I hand-copied your commit with some minor refactoring.
I will leave inline comments on where I modified a bit, but it should almost follow your implementation.

Thanks a lot for your help !

vtjnash commented on base/compiler/abstractinterpretation.jl:
```diff
@@ -788,20 +790,24 @@ end
-        end
-        return tmerge(tx, ty)
+    if f === ifelse && fargs isa Vector{Any} && la == 4 && (cnd = argtypes[2]; isa(cnd, Conditional))
+        if (newcnd = widenconditional(cnd); isa(newcnd, Const))
```
style nit: we try to avoid nesting statements, particularly assignments
```suggestion
        newcnd = widenconditional(cnd)
        if isa(newcnd, Const)
```

vtjnash commented on base/compiler/abstractinterpretation.jl:
```diff
@@ -1277,6 +1303,8 @@ function typeinf_local(interp::AbstractInterpreter, frame::InferenceState)
     s = frame.stmt_types
     n = frame.nstmts
+    nargs = frame.nargs
+    isva = let def = frame.linfo.def; isa(def, Method) && def.isva; end
```
style nit: the `;` is mostly for convenient REPL usage, but we generally try to avoid packing statements on a line (the let block also isn't essential and just adds line noise in this case)
```suggestion
    def = frame.linfo.def
    isva = isa(def, Method) && def.isva
```

aviatesk commented:
@vtjnash @Keno bump, can we push this forward ?

aviatesk committed:
inference: inter-procedural conditional constraint back-propagation

This PR propagates `Conditional`s inter-procedurally when a
`Conditional` at return site imposes a constraint on the call arguments.
When inference exits local frame and the return type is annotated as
`Conditional`, it will be converted into `InterConditional` object,
which is implemented in `Core` and can be directly put into the global
cache. Finally after going back to caller frame, `InterConditional` will
be re-converted into `Conditional` in the context of the caller frame.

 ## improvements

So now some simple "is-wrapper" functions will propagate its constraint
as expected, e.g.:
```julia
isaint(a) = isa(a, Int)
@test Base.return_types((Any,)) do a
    isaint(a) && return a # a::Int
    return 0
end == Any[Int]

isaint2(::Any) = false
isaint2(::Int) = true
@test Base.return_types((Any,)) do a
    isaint2(a) && return a # a::Int
    return 0
end == Any[Int]

function isa_int_or_float64(a)
    isa(a, Int) && return true
    isa(a, Float64) && return true
    return false
end
@test Base.return_types((Any,)) do a
    isa_int_or_float64(a) && return a # a::Union{Float64,Int}
    0
end == Any[Union{Float64,Int}]
```

(and now we don't need something like #38636)

 ## benchmarks

A compile time comparison:
> on the current master (82d79ce18f88923c14d322b70699da43a72e6b32)
```
Sysimage built. Summary:
Total ───────  55.295376 seconds
Base: ───────  23.359226 seconds 42.2444%
Stdlibs: ────  31.934773 seconds 57.7531%
    JULIA usr/lib/julia/sys-o.a
Generating REPL precompile statements... 29/29
Executing precompile statements... 1283/1283
Precompilation complete. Summary:
Total ───────  91.129162 seconds
Generation ──  68.800937 seconds 75.4983%
Execution ───  22.328225 seconds 24.5017%
    LINK usr/lib/julia/sys.dylib
```

> on this PR (37e279bce7136e48f159811641b68143412c3881)
```
Sysimage built. Summary:
Total ───────  51.694730 seconds
Base: ───────  21.943914 seconds 42.449%
Stdlibs: ────  29.748987 seconds 57.5474%
    JULIA usr/lib/julia/sys-o.a
Generating REPL precompile statements... 29/29
Executing precompile statements... 1357/1357
Precompilation complete. Summary:
Total ───────  88.956226 seconds
Generation ──  67.077710 seconds 75.4053%
Execution ───  21.878515 seconds 24.5947%
    LINK usr/lib/julia/sys.dylib
```

Here is a sample code that benefits from this PR:
```julia
function summer(ary)
    r = 0
    for a in ary
        if ispositive(a)
            r += a
        end
    end
    r
end

ispositive(a) = isa(a, Int) && a > 0

ary = Any[]
for _ in 1:100_000
    if rand(Bool)
        push!(ary, rand(-100:100))
    elseif rand(Bool)
        push!(ary, rand('a':'z'))
    else
        push!(ary, nothing)
    end
end

using BenchmarkTools
@btime summer($(ary))
```

> on the current master (82d79ce18f88923c14d322b70699da43a72e6b32)
```
❯ julia summer.jl
  1.214 ms (24923 allocations: 389.42 KiB)
```

> on this PR (37e279bce7136e48f159811641b68143412c3881)
```
❯ julia summer.jl
  421.223 μs (0 allocations: 0 bytes)
```

 ## caveats

Within the `Conditional`/`InterConditional` framework, only a single
constraint can be back-propagated inter-procedurally. This PR implements
a naive heuristic to "pick up" a constraint to be propagated when a
return type is a boolean. The heuristic may fail to select an
"interesting" constraint in some cases. For example, we may expect
`::Expr` constraint to be imposed on the first argument of
`Meta.isexpr`, but the current heuristic ends up picking up a constraint
on the second argument (i.e. `ex.head === head`).
```julia
isexpr(@nospecialize(ex), head::Symbol) = isa(ex, Expr) && ex.head === head

@test_broken Base.return_types((Any,)) do x
    Meta.isexpr(x, :call) && return x # x::Expr, ideally
    return nothing
end == Any[Union{Nothing,Expr}]
```

I think We can get rid of this limitation by extending `Conditional` and
`InterConditional`
so that they can convey multiple constraints, but I'd like to leave this
as a future work.

---

- closes #38636
- closes #37342

aviatesk committed:
tweaks for inter-procedural constraint back-prop'

note that the changes for `isnothing` and `ismissing` aren't necessary,
but they reduce the number of method definitions for good reason;
the less the number of methods they have, the better we can
back-propagate type constraints, because even after a package defines
their own new methods for them we can keep to use our `InterConditional`
logic as far as far as the number of methods is
[≤3](https://github.com/JuliaLang/julia/blob/5c6e21edbfd8f0c7d16ea01c91d1c75c30d4eaa1/base/compiler/types.jl#L119)

aviatesk commented on test/compiler/inference.jl:
```diff
@@ -1727,6 +1729,90 @@ for expr25261 in opt25261[i:end]
+        return nothing
+    end == Any[Union{Nothing,Expr}]
+
+    # TODO: back-propagate multiple conditional constraints
+    @test_broken (@evaltoplevel begin
+        is_int_and_int(a, b) = isa(a, Int) && isa(b, Int)
+        Base.return_types((Any,Any)) do a, b
+            is_int_and_int(a, b) && return a, b # (a::Int, b::Int) ideally, but (a::Any, b::Int)
+            0, 0
+        end
+    end) == Any[Tuple{Int,Int}]
```
I changed the old `isexpr` test case to this one for the future multiple conditional constraint back-propagation enhancement.

aviatesk commented on base/compiler/abstractinterpretation.jl:
```diff
@@ -154,6 +161,26 @@ function abstract_call_gf_by_type(interp::AbstractInterpreter, @nospecialize(f),
         seen += 1
         rettype = tmerge(rettype, this_rt)
         rettype === Any && break
+        if this_conditional !== Bottom && is_lattice_bool(rettype) && !method.isva && fargs !== nothing
```
The `t !== Bottom && t ⊑ Bool` pattern is now wrapped in `is_lattice_bool` check.

aviatesk commented on base/essentials.jl:
```diff
@@ -813,8 +813,7 @@ const missing = Missing()
 """
-ismissing(::Any) = false
-ismissing(::Missing) = true
+ismissing(x) = x === missing
```
@vtjnash should I revert back to the previous `ismissing`/`isnothing` definitions ?
I believe the less method definitions the better, because we can keep to use `InterConditional` logic as far as there're ≦3 methods even after a package defines new methods for them.

xref: https://github.com/JuliaLang/julia/pull/38905/files#r552878124

aviatesk committed:
inference: callsite conditional argument type refinement

- within a callee (i.e. `typeinf_local`), we widen conditional return
  type if it doesn't refine input argument type (for better cache)
- within a caller (i.e. `abstract_call_gf_by_type`), we re-form a
  conditional if needed, which allows us to choose a propagation target
  more appropriately

this commit implements the "pick up" logic within a caller
(i.e. within `abstract_call_gf_by_type`), which allows us to choose a
constraint more appropriately, and now the `Meta.isexpr` case is fixed.
Still there is a limitation of multiple conditional constraint
back-propagation; the following broken test case will explain the future
work.
```julia
is_int_and_int(a, b) = isa(a, Int) && isa(b, Int)
@test_broken Base.return_types((Any,Any)) do a, b
    is_int_and_int(a, b) && return a, b # (a::Int, b::Int) ideally, but (a::Any, b::Int)
    0, 0
end == Any[Tuple{Int,Int}]
```

KristofferC commented:
Probably good to run the benchmarks here before merging.

aviatesk commented:
> Probably good to run the benchmarks here before merging.

Which kind of benchmarks you want ?

KristofferC commented:
I just meant the standard ones.

@nanosoldier `runbenchmarks(ALL, vs = ":master")`

aviatesk commented on base/compiler/abstractinterpretation.jl:
```diff
@@ -164,13 +189,47 @@ function abstract_call_gf_by_type(interp::AbstractInterpreter, @nospecialize(f),
+                    if const_rettype isa LimitedAccuracy
+                        rettype = LimitedAccuracy(rettype, const_rettype.causes)
+                    end
+                end
```
Oh, previous `rettype = Conditional(SlotNumber(const_conditional.slot), const_conditional.vtype, const_conditional.elsetype)` was just wrong. Thanks for the fix.

aviatesk reviewed (commented):
The commits @vtjnash added look great to me. Seems we're good to go ?

vtjnash commented:
The Nanosoldier machine has been down unfortunately, so we can't check that right now. But this is good to merge, so could you clean up and squash the fixup commits so that every commit is correct?

aviatesk commented:
sure, commits squashed into reasonable pieces, and this PR got rebased against the current `master`.

aviatesk commented:
@vtjnash bump. I want to finish this PR and push forward another PRs like #39305 .

nanosoldier commented:
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/38905#issuecomment-765281357):
```
ProcessExitedException(3)
```
cc @christopher-dG

aviatesk commented:
I rebased this PR to take in #39512, and I'm facing the following error when creating a system image.
It seems that some C code modified in the PR breaks the serialization (I confirmed I could build this PR successfully with reverting #39512 ).
@Keno @vtjnash do you have any idea to fix this ? I'm not so familiar with C code as of now.
```
Stdlibs: ────  32.658716 seconds 54.2971%

signal (11): Segmentation fault: 11
in expression starting at none:0
jl_serialize_value_ at /Users/aviatesk/julia/julia/src/staticdata.c:414
jl_save_system_image_to_stream at /Users/aviatesk/julia/julia/src/staticdata.c:1538
jl_save_system_image at /Users/aviatesk/julia/julia/src/staticdata.c:1653
jl_write_compiler_output at /Users/aviatesk/julia/julia/src/precompile.c:81
jl_atexit_hook at /Users/aviatesk/julia/julia/src/init.c:211
repl_entrypoint at /Users/aviatesk/julia/julia/src/jlapi.c:703
Allocations: 113758935 (Pool: 113726722; Big: 32213); GC: 149
```

vtjnash commented:
You need to update NUM_TAGS when you add a tag https://github.com/JuliaLang/julia/pull/38905/files#diff-0e5a5cdf744ddbc1c815bf0a767bc9988e37d88792e8acc95c7ed9863491f2afR33

aviatesk commented:
Woa that's just my mistake. Thanks for pointing out !

vtjnash commented:
Let's see if it is alive: @nanosoldier `runbenchmarks(ALL, vs = ":master")`

aviatesk commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

vtjnash commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/38905#issuecomment-783544483):
```
ProcessExitedException(5)
```
cc @christopher-dG

vtjnash commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/38905#issuecomment-786150438) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/f3babb2_vs_95a34a9/report.md). cc @christopher-dG

Keno commented:
Looks good

aviatesk commented:
To me there seems to be regressions in `sort!` benchmarks ?
(this is the first time for me to look at an nanosoldier benchmark result and so I'm not sure if "40% regression" is something we can allow to happen, though)

Keno commented:
It's running on a new machine that hasn't been fully tuned yet, so as long as there's nothing egregious (which I didn't see) it should be fine.

vtjnash commented:
I've noticed that test can get into some strange bi-stable modes. As long as all of `sort!` tests in a group aren't showing a trend, it is probably nothing.

LGTM

aviatesk commented:
I'm very happy we can merge this at last ! I really appreciate your help @vtjnash and @Keno .
